File: 907.Koko Eating Bananas/solution.cpp
class Solution {
public:
    bool helper(vector<int>& piles, int rate, int h){
        int need = 0;
        for(int i: piles) {
            need = need + (i + rate - 1) / rate;
            if(need > h){
                return false;
            }
        }
        return need <= h;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1;
        int r = 0;
        for(int i:piles) {
            if(r < i) {
                r = i;
            }
        }

        while(l < r) {
            int mid = l + (r - l) / 2;
            if(helper(piles, mid, h)) {
                r = mid;
            }else {
                l = mid + 1;
            }
        }
        return r;
    }
};

File: 907.Koko Eating Bananas/solution.java
class Solution {
    public boolean caneat(int[] piles,int h, int n){
        long hours = 0;
        for(int i:piles){
            hours+=(long)(i+n-1)/n;
            if(hours>h) return false;
        }
        return hours<=h;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = 0;
        for(int i:piles){
            if(right<=i){
                right = i;
            }
        }
        int res = right;

        while(left<=right){
            int mid = left+(right-left)/2;
            if(caneat(piles,h,mid)){
                res = mid;
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return res;
    }
}